### ISSUE TITLE
ARDUINO_VARIANT cannot be cast to (char*) when compiling in Eclipse/windows

### ISSUE THREAD

**Issue by Bascy** [espressif/arduino-esp32#2193]
### Hardware:
Board:							ESP32 Dev Module
Core Installation/update date:		commit ec3a97b598f2a8516297fee873ce78492f5d7e64 (dec 15)
IDE name:						Eclipse / Sloeber 4.3 plugin
Computer OS:                                         Windows 10

### Description:
On compilation of the ESPmDNS component the following code results in an error 
` error: 'esp32' was not declared in this scope`
because the macro ARDUINO_VARIANT is replaced by esp32 and not by the string "esp32"

```
void MDNSResponder::enableArduino(uint16_t port, bool auth){
    mdns_txt_item_t arduTxtData[4] = {
        {(char*)"board"         ,(char*)ARDUINO_VARIANT},
        {(char*)"tcp_check"     ,(char*)"no"},
        {(char*)"ssh_upload"    ,(char*)"no"},
        {(char*)"auth_upload"   ,(char*)"no"}
    };

```

**Comment by Bascy** [comment]
Guess this should be solved as described here:
https://www.guyrutenberg.com/2008/12/20/expanding-macros-into-string-constants-in-c/

**Comment by atanisoft** [comment]
This is more likely an issue with how you are compiling in Eclipse. In Arduino IDE (https://github.com/espressif/arduino-esp32/blob/master/platform.txt#L68) and PIO (https://github.com/espressif/arduino-esp32/blob/master/tools/platformio-build.py#L94) it is passed in as a quoted string.

So, how are you building in Eclipse?

**Comment by Bascy** [comment]
Using Eclipse plugin from Sloeber, The following definition of ARDUINO_VARIANT is used:

`-DARDUINO_VARIANT="esp32"  `

I'm guessing that Eclipse / windows is loosing the double quotes somewhere

**Comment by lbernstone** [comment]
Can you add something about eclipse env quoting to the issue title so the next person finds this?

**Comment by atanisoft** [comment]
@Bascy the difference is how the quotes are declared, can you try:
```
-DARDUINO_VARIANT="\"esp32\""
```
The same issue could also be seen with ARUDINO_BOARD as well if you dig into it further.

**Comment by Bascy** [comment]
Yes this also solves the problem for me, but as i understand it the problem only occurs when compiling on Eclipse and/or Windows OS. 
So I make a pull request with a solution to this, implemented only when using Eclipse

**Comment by atanisoft** [comment]
As mentioned on your other issue, I think these are issues you should be opening against Eclipse Sloeber and not arduino-esp32. The PR you opened may be a solution for this issue *BUT* there will likely be many more issues as you continue adventuring down the path of an IDE that isn't very well documented or used extensively.

**Comment by Bascy** [comment]
I've been using this IDE for over a year now with different boards and havent had this problem before. I'm guessing this is the only place in the arduino-esp32 core or any other cores that i've been using that casts the result of a macro expansion of a commandline -D define into a (char*)

**Comment by atanisoft** [comment]
I used to use Eclipse Sloeber but got fed up with lack of support and moved over to Platform IO which is considerably faster and more stable for me.

But for this issue, I suspect this is just the first of a few areas that use these defines.

Did you create the makefile you are building with or was it generated by Sloeber? If it was done by Sloeber you should probably open an issue against it so they can ensure this gets passed in correctly for any other constants.

**Comment by Bascy** [comment]
It is generated by Sloeber but I think that it is Windows that removes the quotes as it passes the parameter value to the executable.
Did you use Sloeber on Windows or on a different OS?

**Comment by atanisoft** [comment]
Since it was generated by Sloeber, I'd guess they are not generating it correctly. The quotes are considered optional on all platforms and can be stripped by the shell (even on linux). I'd still suggest open a bug for Sloeber so they can work on a proper fix since they are breaking it with their template.

As for when I used Sloeber, I used it on windows for the ESP8266 and Arduino. I use Eclipse itself for other languages and targets but no longer for Arduino/ESP8266 and never for ESP32 (it wasn't supported at the time).
